README.markdown

```diff
@@ -190,49 +190,73 @@ let colour = "red"
+
 
 ### Protocol Conformance
 
-In particular, when adding protocol conformance to a model, prefer adding a separate extension for the protocol methods. This keeps the related methods grouped together with the protocol and can simplify instructions to add a protocol to a class with its associated methods.
-<!-- [@skarol] - this example is bad. I do agree with Goranche - we should separate DataSources and Delegates to separate classes. I know there is a lot of places in code where it is used like this, so we should point a direction here. -->
+When adding protocol conformance to a model, prefer adding a separate extension for the protocol methods. This keeps the related methods grouped together with the protocol and can simplify instructions to add a protocol to a class with its associated methods.
+
+Since the compiler does not allow you to re-declare protocol conformance in a derived class, it is not always required to replicate the extension groups of the base class. This is especially true if the derived class is a terminal class and a small number of methods are being overridden. When to preserve the extension groups is left to the discretion of the author.
+
+For DataSource, Delegate, etc conformances try to separate classes wherever it is possible and keep them out of view controller. If this is not possible put protocol conformances into extensions.
+
+For UIKit view controllers, consider grouping lifecycle, custom accessors, and IBAction with `// MARK:`.
 
 **Preferred**:
 ```swift
-class MyViewController: UIViewController {
-  // class stuff here
+class ViewController: UIViewController {
+  let dataSource: DataSource
+  let scrollDelegate: ScrollDelegate
+  // ...
+}
+
+class DataSource: UICollectionViewDataSource 
+  // ...
 }
 
-// MARK: - UITableViewDataSource
-extension MyViewController: UITableViewDataSource {
-  // table view data source methods
+// MARK: - UICollectionViewDataSource conformance
+extension DataSource: UICollectionViewDataSource {
+  // UICollectionViewDataSource methods
 }
 
-// MARK: - UIScrollViewDelegate
-extension MyViewController: UIScrollViewDelegate {
-  // scroll view delegate methods
+class ScrollDelegate {
+
+}
+// MARK: UIScrollViewDelegate conformance
+class ScrollDelegate: UIScrollViewDelegate {
+  // UIScrollViewDelegate methods
 }
 ```
+**Less Preferred**:
+```swift
+class ViewController: UIViewController {
+  // ...
+}
 
+// MARK: - UICollectionViewDataSource conformance
+extension ViewController: UICollectionViewDataSource {
+  // ...
+}
+
+// MARK: - UIScrollViewDelegate conformance
+extension ViewController: UIScrollViewDelegate {
+  // ...
+}
+```
 **Not Preferred**:
 ```swift
-class MyViewController: UIViewController, UITableViewDataSource, UIScrollViewDelegate {
+class ViewController: UIViewController, UIScrollViewDelegate, UICollectionViewDataSource {
```

Comment:
I am also thinking to add two words about inheritance over `UITableViewController` and `UICollectionViewController`. I've seen this in the codebase, but I consider this a bad practice in general. As you are not able to extract `DataSource` and `Delegate`. Maybe also mention that we prefered (I do, don't know about you @TimPapler yet) composition over inheritance. Imho it is a more flexible approach for code reusability. Let me know what do you think.

Reply:
Yes I prefer composition over inheritance as well. And good point about `UICollection/TableViewController` inheritance .

Reply:
would this be enough: 
> Avoid inheriting from `UICollectionvViewController` and `UITableViewController` use `UIViewController` with composition instead.

Reply:
yes. I think this will work